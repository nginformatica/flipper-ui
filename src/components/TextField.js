import React from 'react'
import _ from 'prop-types'
import styled from 'styled-components'
import { background, primary, text } from '../colors'

const StyledTextField = styled.input`
    outline: none;
    padding: 8px;
    border-radius: 6px;
    line-height: 1.5em;
    margin: 1em;
    border: 1px solid ${background.dark};
    color: ${text};
    font-size: 14px;
    font-family: 'Roboto', sans-serif;
    transition: all 500ms ease;
    height: 24px;
    &:focus {
        border: 1px solid ${primary.normal};
        border-radius: 6px;
    }
`

const TextField = props => <StyledTextField { ...props } />

TextField.propTypes = {
    placeholder: _.string,
    value: _.string,
    style: _.object,
    onChange: _.func
}

export default TextField
