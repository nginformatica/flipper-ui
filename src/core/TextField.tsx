import {
    InputAdornment,
    TextField as MuiTextField,
    IconButton as MuiButton
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import React, {
    ChangeEvent,
    KeyboardEvent,
    FC,
    FocusEvent,
    ReactNode,
    MouseEvent
} from 'react'
import { DefaultProps } from './types'
import { Clear, Help as ContactSupportIcon } from '@material-ui/icons'
import IconButton from './IconButton'
import styled from 'styled-components'

export interface TextFieldProps extends DefaultProps {
    autoComplete?: string
    autoFocus?: boolean
    defaultValue?: string | number
    disabled?: boolean
    error?: boolean
    fullWidth?: boolean
    id?: string
    label?: string
    placeholder?: string
    multiline?: boolean
    name?: string
    required?: boolean
    select?: boolean
    type?: string
    value?: string | number
    variant?: 'standard' | 'outlined' | 'filled'
    inputRef?: React.Ref<HTMLInputElement>
    inputProps?: object
    InputProps?: object
    InputLabelProps?: object
    SelectProps?: object
    rows?: string | number
    rowsMax?: string | number
    helperText?: ReactNode
    helperIcon?: ReactNode
    hasClear?: boolean
    onClear?: () => void
    onHelperClick?: () => void
    onChange?: (event: ChangeEvent<HTMLInputElement>) => void
    onPaste?: (event: React.ClipboardEvent<HTMLInputElement>) => void
    onBlur?: (event: FocusEvent<HTMLInputElement>) => void
    onKeyUp?: (event: KeyboardEvent) => void
    onKeyDown?: (event: KeyboardEvent) => void
}

interface IHelperProps extends Pick<TextFieldProps, 'helperIcon'> {
    onHelperClick: (event: MouseEvent<HTMLButtonElement>) => void
}

const Helper = styled.div`
    width: 42px;
    height: 38px;
`

export const TextFieldWrapper = styled.div`
    display: flex;
    flex-direction: rows;
    width: 100%;
    align-items: center;
    button {
        display: none;
    }
    :hover button {
        display: flex;
    }
`

export const StaticTextFieldWrapper = styled.div`
    display: flex;
    flex-direction: rows;
    width: 100%;
    align-items: center;
`

export const useStyles = makeStyles({
    input: {
        fontSize: '14px',
        padding: '10px',
        height: 'auto'
    },
    outlinedInput: {
        fontSize: '14px',
        padding: '10px',
        height: 'auto'
    },
    outlinedLabel: {
        fontSize: '14px',
        transform: 'translate(14px, 13px) scale(1)'
    },
    outlinedMultiline: {
        padding: '0px'
    },
    iconOutlined: {
        right: '2px'
    }
})

export const HelperBox = (props: IHelperProps) => (
    <Helper>
        <IconButton padding='6px 2px' onClick={props.onHelperClick}>
            {props.helperIcon || <ContactSupportIcon color='primary' />}
        </IconButton>
    </Helper>
)

const renderEndAdornment = (onClear?: () => void) => (
    <InputAdornment position='end'>
        <MuiButton onClick={onClear} size='small'>
            <Clear style={{ fontSize: '15px' }} />
        </MuiButton>
    </InputAdornment>
)

const TextField: FC<TextFieldProps> = ({
    margin,
    padding,
    style = {},
    error,
    variant = 'outlined',
    InputLabelProps = {},
    InputProps = {},
    SelectProps = {},
    autoComplete = 'off',
    onHelperClick,
    helperIcon,
    fullWidth,
    hasClear,
    onClear,
    ...otherProps
}) => {
    const clearStyle = makeStyles({
        iconOutlined: {
            position: 'relative',
            right: '2px',
            marginLeft: '-20px'
        }
    })

    const hasValue = !!otherProps.value

    const endAdornment =
        hasValue && hasClear
            ? { endAdornment: renderEndAdornment(onClear) }
            : {}

    const classes = useStyles()
    const clearClass = clearStyle()

    const handleClick = () => {
        if (onHelperClick) {
            onHelperClick()
        }
    }

    const Wrapper = hasClear ? StaticTextFieldWrapper : TextFieldWrapper

    return (
        <Wrapper>
            <MuiTextField
                fullWidth={fullWidth}
                autoComplete={autoComplete}
                error={error}
                variant={variant as 'outlined'}
                style={{
                    margin,
                    padding,
                    ...style
                }}
                InputLabelProps={{
                    classes: {
                        outlined:
                            variant === 'outlined' ? classes.outlinedLabel : ''
                    },
                    ...InputLabelProps
                }}
                InputProps={{
                    classes: {
                        input:
                            variant === 'outlined' ? classes.outlinedInput : '',
                        multiline:
                            variant === 'outlined'
                                ? classes.outlinedMultiline
                                : ''
                    },
                    ...InputProps
                }}
                SelectProps={{
                    classes: {
                        iconOutlined: hasClear
                            ? clearClass.iconOutlined
                            : classes.iconOutlined
                    },
                    ...endAdornment,
                    ...SelectProps
                }}
                {...otherProps}
            />
            {onHelperClick && (
                <HelperBox
                    helperIcon={helperIcon}
                    onHelperClick={handleClick}
                />
            )}
        </Wrapper>
    )
}

export default TextField
