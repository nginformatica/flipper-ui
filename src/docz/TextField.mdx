---
name: TextField
route: /text-field
---

import InputAdornment from '../core/InputAdornment'
import TextField from '../core/TextField'
import ListItem from '../core/ListItem'
import { Playground, Props } from 'docz'

# TextField

## Properties
<Props of={ TextField } />

## Simple TextField
<Playground>
    <TextField placeholder='Description' />
</Playground>

## TextField with InputAdornment
<Playground>
    <TextField
        type='number'
        label='Price'
        InputProps={ {
            startAdornment: <InputAdornment position='start'>$</InputAdornment>
        } }
        InputLabelProps={ { shrink: true } }
    />
</Playground>


## Simple Text field with helper button
<Playground>
    <TextField 
        placeholder='Description' 
        onHelperClick={ () => window.alert('HELP!')} 
    />
</Playground>

## TextField select
<Playground>
    <div>
        <TextField select value='reais'>
            {
                [
                    { label: 'R$', value: 'reais' },
                    { label: '$', value: 'dollar' }
                ].map(({ label, value }) =>
                    <ListItem
                        key={ value }
                        value={ value }>
                        { label }
                    </ListItem>
                )
            }
        </TextField>
        <TextField select value='functional-frontend' value='fable'>
            {
                [
                    { label: 'Elm', value: 'elm' },
                    { label: 'ReasonML', value: 'reasonml' },
                    { label: 'Purescript', value: 'purescript' },
                    { label: 'Fable', value: 'fable' }
                ].map(({ label, value }) =>
                    <ListItem
                        key={ value }
                        value={ value }>
                        { label }
                    </ListItem>
                )
            }
        </TextField>
    </div>
</Playground>
