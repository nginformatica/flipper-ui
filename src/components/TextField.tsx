import {
    createStyles,
    WithStyles,
    withStyles
} from '@material-ui/core'
import { TextField as MuiTextField } from '@material-ui/core'
import { merge } from 'ramda'
import React from 'react'
import { background, primary } from '../colors'

interface IProps extends WithStyles<typeof styles> {
    autoComplete?: string
    autoFocus?: boolean
    defaultValue?: string | number
    disabled?: boolean
    error?: boolean
    fullWidth?: boolean
    id?: string
    label?: string
    placeholder?: string
    style?: object
    margin?: 'none' | 'dense' | 'normal'
    multiline?: boolean
    name?: string
    required?: boolean
    select?: boolean
    type?: string
    value?: string | number | boolean | string[]
    InputProps?: object
    InputLabelProps?: object
    SelectProps?: object
    onChange?: (event) => {}
}

const styles = theme => createStyles({
    input: {
        '&:focus': {
            borderColor: theme.palette.primary.light,
            boxShadow: `0 0 0 0.2rem ${theme.palette.primary.light}D9`,
        },
        'backgroundColor': theme.palette.common.white,
        'border': `1px solid ${background.dark}`,
        'borderRadius': 4,
        'fontSize': 16,
        'padding': '10px 12px',
        'transition': theme.transitions.create(['border-color', 'box-shadow']),
        'width': 'calc(100% - 24px)'
    },
    label: {
        fontSize: 18
    },
    root: {
        'label + &': {
            marginTop: theme.spacing.unit * 3,
        },
        'padding': 0
    }
})

const TextField = ({ classes, InputProps, InputLabelProps, ...otherProps }: IProps) =>
    <MuiTextField
        { ...otherProps }
        InputProps={
            merge(
                {
                    classes: {
                        input: classes.input,
                        root: classes.root
                    },
                    disableUnderline: true
                },
                InputProps
            )
        }
        InputLabelProps={
            merge({
                    className: classes.label,
                    shrink: true
                },
                InputLabelProps
            )
        }
    />

export default withStyles(styles)(TextField)
