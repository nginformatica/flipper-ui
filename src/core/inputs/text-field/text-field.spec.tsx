import * as React from 'react'
import { act } from 'react-dom/test-utils'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import TextField from '@/test/mocks/text-field-mock'
import TextFieldOptions from '@/test/mocks/text-field-options-mock'

const LIST = [
    { label: 'Elm', value: 'elm' },
    { label: 'ReasonML', value: 'reasonml' },
    { label: 'Purescript', value: 'purescript' },
    { label: 'Fable', value: 'fable' }
]

describe('TextField', () => {
    it('should render', () => {
        render(<TextField inputProps={{ placeholder: 'Description' }} />)
        const textField = screen.getByPlaceholderText('Description')

        expect(textField).toBeDefined()
    })

    it('should render outlined', () => {
        render(
            <TextField
                inputProps={{ placeholder: 'Description', variant: 'outlined' }}
            />
        )
        const textField = screen.getByPlaceholderText('Description')

        expect(textField.classList).toContain('MuiOutlinedInput-input')
    })

    it('should render filled', () => {
        render(
            <TextField
                inputProps={{ placeholder: 'Description', variant: 'filled' }}
            />
        )
        const textField = screen.getByPlaceholderText('Description')

        expect(textField.classList).toContain('MuiInputBase-input')
    })

    it('should render with helper', async () => {
        const onHelperClick = jest.fn()

        render(
            <TextField
                inputProps={{
                    placeholder: 'Description',
                    onHelperClick: onHelperClick
                }}
            />
        )
        const textField = screen.getByPlaceholderText('Description')

        await userEvent.hover(textField)

        const helperBox = screen.getByRole('helper-box')

        await act(
            async () =>
                await userEvent.click(helperBox.firstElementChild || helperBox)
        )

        expect(onHelperClick).toBeCalled()
    })

    it('should render select with clear button', async () => {
        render(
            <TextField
                initialOption='elm'
                inputProps={{
                    select: true,
                    placeholder: 'Description',
                    hasClear: true
                }}
            />
        )

        const textField = screen.getByPlaceholderText(
            'Description'
        ) as HTMLInputElement

        expect(textField.value).toBe('elm')

        const clearBtn = screen.getByRole('clear-button')

        await act(async () => await userEvent.click(clearBtn))

        expect(textField.value).toBe('')
    })

    it('should update on type', () => {
        render(<TextField inputProps={{ placeholder: 'Description' }} />)
        const textField = screen.getByPlaceholderText(
            'Description'
        ) as HTMLInputElement

        fireEvent.change(textField, { target: { value: 'Hello' } })

        expect(textField.value).toBe('Hello')
    })

    it('should update on type', () => {
        render(
            <TextFieldOptions
                inputProps={{ placeholder: 'Description' }}
                options={LIST}
            />
        )
        const textField = screen.getByPlaceholderText(
            'Description'
        ) as HTMLInputElement

        fireEvent.change(textField, { target: { value: '' } })

        expect(textField.value).toBe('')
    })

    it('should update on type', () => {
        render(
            <TextFieldOptions
                inputProps={{ placeholder: 'Description' }}
                options={JSON.stringify(LIST)}
            />
        )
        const textField = screen.getByPlaceholderText(
            'Description'
        ) as HTMLInputElement

        fireEvent.change(textField, { target: { value: '' } })

        expect(textField.value).toBe('')
    })

    it.only('should render character counter', () => {
        render(
            <TextField characters inputProps={{ placeholder: 'Description' }} />
        )
        const counter = screen.getByTestId('characters-counter')

        waitFor(() => {
            expect(counter).toBeDefined()
        })
    })
})
