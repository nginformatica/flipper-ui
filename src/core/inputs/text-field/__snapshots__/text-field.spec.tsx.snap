// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should match snapshot 1`] = `
<div>
  <div
    class="sc-fqkwJk iUjhJF"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      title=""
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-controls=":re:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":rd:"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input makeStyles-input-41 MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-18dp293-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id=":rd:"
          role="combobox"
          tabindex="0"
        >
          Elm
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          autocomplete="off"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          placeholder="Description"
          tabindex="-1"
          value="elm"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
            role="clear-button"
            style="margin: 0px 16px 0px 0px;"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="ClearIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
